buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java

```diff
@@ -0,0 +1,302 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle.testclusters;
+
+import org.elasticsearch.GradleServicesAdapter;
+import org.elasticsearch.gradle.Distribution;
+import org.elasticsearch.gradle.Version;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class ElasticsearchCluster implements TestClusterConfiguration {
+
+    private static final int CLUSTER_UP_TIMEOUT = 40;
+    private static final TimeUnit CLUSTER_UP_TIMEOUT_UNIT = TimeUnit.SECONDS;
+
+    private final AtomicBoolean configurationFrozen = new AtomicBoolean(false);
+    private final Logger logger = Logging.getLogger(ElasticsearchNode.class);
+    private final String path;
+    private final String clusterName;
+    private final NamedDomainObjectContainer<ElasticsearchNode> nodes;
+    private final File workingDirBase;
+    private final LinkedHashMap<String, Predicate<TestClusterConfiguration>> waitConditions = new LinkedHashMap<>();
+
+    public ElasticsearchCluster(String path, String clusterName, Project project, File artifactsExtractDir, File workingDirBase) {
+        this.path = path;
+        this.clusterName = clusterName;
+        nodes = project.container(ElasticsearchNode.class);
+        this.nodes.add(
+            new ElasticsearchNode(
+                path, clusterName + "-1",
+                GradleServicesAdapter.getInstance(project), artifactsExtractDir, workingDirBase
+            )
+        );
+        this.workingDirBase = workingDirBase;
+    }
+
+    public void setNumberOfNodes(int numberOfNodes) {
+        checkFrozen();
+        if (numberOfNodes < 1) {
+            throw new IllegalArgumentException("Number of nodes should be >= 1 but was " + numberOfNodes);
+        }
+        if (numberOfNodes <= nodes.size()) {
```

Comment:
I don't see any reason to enforce this constraint. In fact, there's no reason to eagerly create instances of `ElasticsearchNode` at all here, including the initial one in the constructor so far as I can tell. Since this is a `DomainObjectCollection` we can configure it lazily and delay instantiation of the nodes when we need them. This would allow for more flexible configuration, as well as create less garbage when executing builds that never use these clusters.

Reply:
Why would you want to start with a larger cluster and shrink it in a different part of the build ? On the other hand the same argument could be made for growing the cluster, so maybe we should limit any change if  ` nodes.size() != 1`. 

I think that there's a place for settings that are configured only once throughout the build, changing their value from some default to prevent hard to trace back-and-forth configuration in larger projects such as ours. 

I'm not sure about the lazy configuration, granted I'm not super familiar with the Gradle APIs for it, but especially if we do keep the restriction I think there's little benefit and not  worth spending the time now.

Reply:
From a performance perspective it'd be nice to defer creating all these things until we need them but it's probably trivial.

FWIW, `collection.all { }` as used in the plugin already is an example of the lazy configuration APIs I'm referring to.